domain/domain.go

```diff
@@ -4,32 +4,169 @@ import (
 type Domain interface {
 	MiningManager() miningmanager.MiningManager
 	Consensus() externalapi.Consensus
+	TemporaryConsensus() externalapi.Consensus
+	CreateTemporaryConsensus() error
+	CommitTemporaryConsensus() error
 }
-
 type domain struct {
-	miningManager miningmanager.MiningManager
-	consensus     externalapi.Consensus
+	miningManager   miningmanager.MiningManager
+	consensus       *externalapi.Consensus
+	tempConsensus   *externalapi.Consensus // We assume there's no concurrent access to tempConsensus
+	consensusConfig *consensus.Config
+	db              infrastructuredatabase.Database
+}
+
+func (d *domain) Consensus() externalapi.Consensus {
+	return *d.consensus
 }
 
-func (d domain) Consensus() externalapi.Consensus {
-	return d.consensus
+func (d *domain) TemporaryConsensus() externalapi.Consensus {
+	return *d.tempConsensus
 }
 
-func (d domain) MiningManager() miningmanager.MiningManager {
+func (d *domain) MiningManager() miningmanager.MiningManager {
 	return d.miningManager
 }
 
+func (d *domain) CreateTemporaryConsensus() error {
+	_, hasInactivePrefix, err := prefixmanager.InactivePrefix(d.db)
+	if err != nil {
+		return err
+	}
+
+	if hasInactivePrefix {
+		return errors.Errorf("cannot have more than one inactive prefix")
+	}
+
+	activePrefix, exists, err := prefixmanager.ActivePrefix(d.db)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return errors.Errorf("cannot create a temporary consensus when there's " +
+			"no active consensus")
+	}
+
+	var inactivePrefix byte = 0
+	if activePrefix == 0 {
+		inactivePrefix = 1
+	}
+
+	err = prefixmanager.SetPrefixAsInactive(d.db, inactivePrefix)
+	if err != nil {
+		return err
+	}
+
+	consensusFactory := consensus.NewFactory()
+	consensusInstance, err := consensusFactory.NewConsensus(d.consensusConfig, d.db, inactivePrefix)
+	if err != nil {
+		return err
+	}
+
+	d.tempConsensus = &consensusInstance
+	return nil
+}
+
+func (d *domain) CommitTemporaryConsensus() error {
+	transaction, err := d.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer transaction.RollbackUnlessClosed()
+
+	inactivePrefix, hasInactivePrefix, err := prefixmanager.InactivePrefix(d.db)
+	if err != nil {
+		return err
+	}
+
+	if !hasInactivePrefix {
+		return errors.Errorf("there's no inactive prefix to commit")
+	}
+
+	activePrefix, exists, err := prefixmanager.ActivePrefix(transaction)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return errors.Errorf("cannot commit a temporary consensus when there's " +
+			"no active consensus")
+	}
+
+	err = prefixmanager.SetPrefixAsActive(transaction, inactivePrefix)
+	if err != nil {
+		return err
+	}
+
+	err = prefixmanager.SetPrefixAsInactive(transaction, activePrefix)
+	if err != nil {
+		return err
+	}
+
+	err = transaction.Commit()
+	if err != nil {
+		return err
+	}
+
+	// We delete anything associated with the old prefix outside
+	// of the transaction in order to save memory.
+	err = prefixmanager.DeleteInactivePrefix(d.db)
```

Comment:
Also, if we crash in the middle, we have half a prefix.
I think maybe we want in CreateTemporaryConsensus to call DeleteInactivePrefix(P) right before creating P, just in case there's something there.

EDIT: just read that in domain.New you are doing DeleteInactivePrefix, so that might be sufficient. Leaving this comment just to make sure.